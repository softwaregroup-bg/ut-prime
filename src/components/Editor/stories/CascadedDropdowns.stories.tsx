import { userEvent, within } from '@storybook/testing-library';
import React from 'react';
import Editor from '..';
export {default} from '../Editor.stories';

export const CascadedDropdowns = () =>
    <Editor
        schema={{
            type: 'object',
            properties: {
                continent: {widget: {type: 'dropdown', dropdown: 'continent'}},
                country: {widget: { type: 'dropdown', dropdown: 'country', parent: 'continent'}},
                city: {widget: { type: 'dropdown', dropdown: 'city', parent: 'country'}}
            }
        }}
        cards={{
            edit: {
                className: 'xl:col-3',
                widgets: ['continent', 'country', 'city']
            }
        }}
        onDropdown={() => Promise.resolve({
            continent: [
                {value: 1, label: 'Africa'},
                {value: 2, label: 'Asia'},
                {value: 3, label: 'Europe'},
                {value: 4, label: 'North America'},
                {value: 5, label: 'Oceania'},
                {value: 6, label: 'South America'}
            ],
            country: [
                {value: 1, label: 'Kenya', parent: 1},
                {value: 2, label: 'Uganda', parent: 1},
                {value: 3, label: 'Tanzania', parent: 1},
                {value: 7, label: 'United States', parent: 4},
                {value: 9, label: 'Mexico', parent: 4},
                {value: 10, label: 'Australia', parent: 5},
                {value: 11, label: 'New Zealand', parent: 5},
                {value: 12, label: 'France', parent: 3},
                {value: 13, label: 'Germany', parent: 3},
                {value: 14, label: 'Italy', parent: 3},
                {value: 15, label: 'Spain', parent: 3}
            ],
            city: [
                {value: 1, label: 'Nairobi', parent: 1},
                {value: 2, label: 'Kampala', parent: 2},
                {value: 3, label: 'Dar es Salaam', parent: 3},
                {value: 4, label: 'New York', parent: 7},
                {value: 6, label: 'Mexico City', parent: 9},
                {value: 7, label: 'Canberra', parent: 10},
                {value: 8, label: 'Paris', parent: 12},
                {value: 9, label: 'Berlin', parent: 13},
                {value: 10, label: 'Rome', parent: 14}
            ]
        })}
        layouts={{edit: ['edit']}}
        toolbar
    />;

CascadedDropdowns.play = async({canvasElement}) => {
    const canvas = within(canvasElement);
    const body = within(document.body);
    userEvent.click(await canvas.findByTestId('continent'));
    userEvent.click(body.getByText('Europe'));
    userEvent.click(canvas.getByTestId('country'));
    userEvent.click(body.getByText('France'));
    userEvent.click(canvas.getByTestId('city'));
    userEvent.click(body.getByText('Paris'));
};
