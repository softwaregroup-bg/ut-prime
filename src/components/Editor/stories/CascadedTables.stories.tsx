import { userEvent, within } from '@storybook/testing-library';
import React from 'react';
import Editor from '..';
export {default} from '../Editor.stories';

export const CascadedTables = () =>
    <Editor
        id = {1}
        onGet={() => Promise.resolve({
            person: [
                {personId: 1, fullName: 'John Doe', birthDate: '2000-01-08', nationalId: '777777777'},
                {personId: 2, fullName: 'Jane Doe', birthDate: '2002-03-18', nationalId: '888888888'}
            ],
            document: [
                {documentId: 1, personId: 1, documentTypeId: 1, documentType: 'Passport'},
                {documentId: 2, personId: 1, documentTypeId: 2, documentType: 'Driving License'},
                {documentId: 3, personId: 1, documentTypeId: 3, documentType: 'Marriage Certificate'},
                {documentId: 4, personId: 2, documentTypeId: 1, documentType: 'Passport'},
                {documentId: 5, personId: 2, documentTypeId: 2, documentType: 'Driving License'}

            ].map(item => ({...item, issueDate: '2020-01-0' + item.documentId, expiryDate: '2025-01-0' + item.documentId})),
            attachment: [
                {attachmentId: 1, documentId: 1, pageNumber: 1, contentType: 'image/jpeg', sizeBytes: 12345},
                {attachmentId: 1, documentId: 1, pageNumber: 2, contentType: 'image/jpeg', sizeBytes: 23456},
                {attachmentId: 1, documentId: 2, pageNumber: 1, contentType: 'image/png', sizeBytes: 33333},
                {attachmentId: 1, documentId: 2, pageNumber: 2, contentType: 'image/png', sizeBytes: 22222}
            ]
        })}
        onDropdown={names => Promise.resolve({
            documentType: [
                {value: 1, label: 'Passport'},
                {value: 2, label: 'Driving License'},
                {value: 3, label: 'Marriage Certificate'}
            ]
        })}
        cards={{
            person: {
                className: 'xl:col-4',
                label: 'Person',
                widgets: [{
                    name: 'person',
                    hidden: ['personId'],
                    widgets: ['fullName', 'nationalId', 'birthDate']
                }]
            },
            document: {
                className: 'xl:col-4',
                label: 'Document',
                widgets: [{
                    name: 'document',
                    hidden: ['documentId', 'personId'],
                    widgets: ['documentTypeId', 'issueDate', 'expiryDate']
                }]
            },
            attachment: {
                className: 'xl:col-4',
                label: 'Attachment',
                widgets: [{
                    name: 'attachment',
                    hidden: ['attachmentId', 'documentId'],
                    widgets: ['pageNumber', 'contentType', 'sizeBytes']
                }]
            }
        }}
        schema={{
            type: 'object',
            properties: {
                person: {
                    type: 'array',
                    items: {
                        type: 'object',
                        properties: {
                            fullName: {},
                            nationalId: {title: 'National ID'},
                            birthDate: {format: 'date'}
                        }
                    },
                    title: '',
                    widget: {
                        type: 'table',
                        selectionMode: 'single'
                    }
                },
                document: {
                    type: 'array',
                    items: {
                        properties: {
                            documentTypeId: {body: 'documentType', widget: {type: 'dropdown', dropdown: 'documentType'}},
                            issueDate: {format: 'date'},
                            expiryDate: {format: 'date'}
                        }
                    },
                    title: '',
                    widget: {
                        type: 'table',
                        master: {
                            personId: 'personId'
                        },
                        parent: '$.selected.person',
                        autoSelect: true,
                        selectionMode: 'single'
                    }
                },
                attachment: {
                    type: 'array',
                    items: {
                        properties: {
                            pageNumber: {type: 'integer'},
                            contentType: {},
                            sizeBytes: {type: 'integer', title: 'Size (bytes)'}
                        }
                    },
                    title: '',
                    widget: {
                        type: 'table',
                        master: {
                            documentId: 'documentId'
                        },
                        parent: '$.selected.document',
                        autoSelect: true,
                        selectionMode: 'single'
                    }
                }
            }
        }}
        layouts={{edit: ['person', 'document', 'attachment']}}
    />;

CascadedTables.play = async({canvasElement}) => {
    const canvas = within(canvasElement);
    userEvent.click(await canvas.findByText('John Doe'));
    userEvent.click(canvas.getByText('Driving License'));
};
