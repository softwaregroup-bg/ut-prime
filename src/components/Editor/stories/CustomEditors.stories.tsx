import React from 'react';
import Editor from '..';
export {default} from '../Editor.stories';

export const CustomEditors = () => {
    function Period({Input, Label, ErrorLabel}) {
        return (
            <>
                <ErrorLabel/>
                <div className='field grid w-full mx-0'>
                    <Label className='md:col-2' name='period' />
                    <Input fieldClass='md:col-4' name='period' />
                    <Input fieldClass='md:col-6' name='unit' />
                </div>
            </>
        );
    }

    Period.properties = ['period', 'unit'];

    return (<Editor
        id={1}
        onGet={() => Promise.resolve({
            period: 5,
            unit: 'days'
        })}
        onDropdown={names => Promise.resolve({})}
        schema={{
            properties: {
                unit: {
                    widget: {
                        type: 'select',
                        options: [
                            {value: 'minutes', label: 'Minutes'},
                            {value: 'hours', label: 'Hours'},
                            {value: 'days', label: 'Days'},
                            {value: 'months', label: 'Months'}
                        ]
                    }
                },
                period: {
                    title: 'Expiration period',
                    type: 'integer'
                }
            }
        }}
        cards={{
            edit: {
                widgets: ['Period']
            }
        }}
        editors={{
            Period
        }}
    />);
};
