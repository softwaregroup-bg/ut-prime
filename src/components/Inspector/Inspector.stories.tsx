import React from 'react';
import type { Story, Meta } from '@storybook/react';
import merge from 'ut-function.merge';

import page from './README.mdx';
import type { Props } from './Inspector.types';
import Inspector from './index';
import Editor from '../Editor';
import decorators from '../test/decorator';

const meta: Meta = {
    title: 'Inspector',
    component: Inspector,
    parameters: {docs: {page}},
    decorators,
    args: {
        state: {}
    }
};
export default meta;

declare type StoryTemplate = Story<Partial<Props>> & {
    play: (context: {canvasElement: HTMLElement}) => Promise<void>
}

const Template: Story<Props> = args => {
    const [value] = React.useState({type: 'object', properties: {}});
    const [override, setOverride] = React.useState({properties: {test: {sort: true}}});
    return <>
        <Inspector {...args} Editor={Editor} onChange={setOverride} object={override} property='properties.test' className='col-3' />
        <pre>{JSON.stringify(merge({}, value, override), null, 2)}</pre>
    </>;
};

export const Basic: StoryTemplate = Template.bind({});
Basic.args = {};
