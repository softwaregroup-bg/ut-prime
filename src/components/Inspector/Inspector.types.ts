import React from 'react';
import {ComponentProps as Editor} from '../Editor/Editor.types';

export interface Props {
    Editor: Editor;
    className?: string;
    object: object;
    property: string | string[];
    type: 'field' | 'card';
    onChange?: (params: object) => void
}

export type ComponentProps = React.FC<Props>
