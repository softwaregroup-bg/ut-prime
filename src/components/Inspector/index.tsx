import React from 'react';
import clsx from 'clsx';
import {createUseStyles} from 'react-jss';
import lodashSet from 'lodash.set';
import lodashGet from 'lodash.get';
import type {ComponentProps} from './Inspector.types';

import type {Props as EditorProps} from '../Editor/Editor.types';

const useStyles = createUseStyles({
    inspector: {
        '& .p-card-body': {
            padding: '0.25rem',
            '& .p-card-content': {
                paddingTop: 0,
                '& .field': {
                    marginBottom: '0.25rem'
                }
            }
        }
    }
});

const props : Record<string, EditorProps> = {
    schema: {
        schema: {
            type: 'object',
            properties: {
                title: {type: 'string', widget: {clear: 'pi-replay'}},
                body: {type: 'string', widget: {clear: 'pi-replay'}},
                filter: {type: 'boolean', widget: {clear: 'pi-replay'}},
                sort: {type: 'boolean', widget: {clear: 'pi-replay'}},
                type: {
                    type: 'string',
                    widget: {type: 'dropdown', dropdown: 'type', showClear: 'pi-replay'}
                },
                widget: {
                    type: 'object',
                    properties: {
                        type: {
                            type: 'string',
                            title: 'Widget',
                            widget: {type: 'dropdown', dropdown: 'widget', showClear: 'pi-replay'}
                        },
                        labelClass: {
                            type: 'string',
                            title: 'Label Class',
                            default: 'md:col-4',
                            widget: {clear: 'pi-replay', type: 'chips', className: 'h-10rem'}
                        },
                        className: {
                            type: 'string',
                            title: 'Input Class',
                            default: 'w-full',
                            widget: {clear: 'pi-replay', type: 'chips', className: 'h-10rem'}
                        },
                        fieldClass: {
                            type: 'string',
                            title: 'Field Class',
                            default: 'md:col-8',
                            widget: {clear: 'pi-replay', type: 'chips', className: 'h-10rem'}
                        },
                        parent: {
                            type: 'string',
                            widget: {clear: 'pi-replay'}
                        }
                    }
                }
            }
        },
        cards: {
            edit: {
                label: 'Field',
                className: 'w-full p-0',
                classes: {
                    default: {
                        label: 'md:col-3',
                        field: 'md:col-9'
                    }
                },
                widgets: [
                    'title',
                    'body',
                    'filter',
                    'sort',
                    'type',
                    'widget.type',
                    'widget.parent',
                    'widget.labelClass',
                    'widget.fieldClass',
                    'widget.className'
                ]
            }
        },
        onDropdown: () => Promise.resolve({
            type: [
                {value: 'string', label: 'String'},
                {value: 'number', label: 'Number'},
                {value: 'boolean', label: 'Boolean'},
                {value: 'object', label: 'Object'},
                {value: 'array', label: 'Array'}
            ],
            widget: [
                {value: 'autocomplete', label: 'autocomplete'},
                {value: 'boolean', label: 'boolean'},
                {value: 'button', label: 'button'},
                {value: 'currency', label: 'currency'},
                {value: 'date-time', label: 'date-time'},
                {value: 'date', label: 'date'},
                {value: 'dropdown', label: 'dropdown'},
                {value: 'dropdownTree', label: 'dropdownTree'},
                {value: 'file', label: 'file'},
                {value: 'icon', label: 'icon'},
                {value: 'image', label: 'image'},
                {value: 'integer', label: 'integer'},
                {value: 'label', label: 'label'},
                {value: 'mask', label: 'mask'},
                {value: 'multiSelect', label: 'multiSelect'},
                {value: 'multiSelectPanel', label: 'multiSelectPanel'},
                {value: 'multiSelectTree', label: 'multiSelectTree'},
                {value: 'multiSelectTreeTable', label: 'multiSelectTreeTable'},
                {value: 'number', label: 'number'},
                {value: 'password', label: 'password'},
                {value: 'radio', label: 'radio'},
                {value: 'select-table-radio', label: 'select-table-radio'},
                {value: 'select', label: 'select'},
                {value: 'selectTable', label: 'selectTable'},
                {value: 'string', label: 'string'},
                {value: 'table', label: 'table'},
                {value: 'text', label: 'text'},
                {value: 'time', label: 'time'}
            ]
        })
    },
    card: {
        schema: {
            type: 'object',
            properties: {
                label: {type: 'string', widget: {clear: 'pi-replay'}},
                watch: {type: 'string', widget: {clear: 'pi-replay'}},
                match: {type: 'string', widget: {clear: 'pi-replay'}},
                className: {
                    type: 'string',
                    title: 'Class',
                    default: 'xl:col-6',
                    widget: {clear: 'pi-replay', type: 'chips', className: 'h-6rem'}
                },
                type: {
                    type: 'string',
                    widget: {
                        type: 'select',
                        options: [
                            {value: 'card', label: 'Card'},
                            {value: 'toolbar', label: 'Toolbar'}
                        ]
                    }
                },
                hidden: {type: 'boolean', widget: {clear: 'pi-replay'}},
                classes: {
                    type: 'object',
                    properties: {
                        card: {
                            type: 'string',
                            title: 'Card class',
                            widget: {clear: 'pi-replay', type: 'chips', className: 'h-8rem'}
                        },
                        default: {
                            type: 'object',
                            properties: {
                                widget: {
                                    type: 'string',
                                    title: 'Widget class',
                                    default: 'field grid',
                                    widget: {clear: 'pi-replay', type: 'chips', className: 'h-8rem'}
                                },
                                label: {
                                    type: 'string',
                                    title: 'Label class',
                                    default: 'md:col-4',
                                    widget: {clear: 'pi-replay', type: 'chips', className: 'h-8rem'}
                                },
                                field: {
                                    type: 'string',
                                    title: 'Field class',
                                    default: 'md:col-8',
                                    widget: {clear: 'pi-replay', type: 'chips', className: 'h-8rem'}
                                },
                                input: {
                                    type: 'string',
                                    title: 'Input class',
                                    default: 'w-full',
                                    widget: {clear: 'pi-replay', type: 'chips', className: 'h-8rem'}
                                }
                            }
                        }
                    }
                }
            }
        },
        cards: {
            edit: {
                label: 'Card',
                className: 'w-full p-0',
                classes: {
                    default: {
                        label: 'md:col-3',
                        field: 'md:col-9'
                    }
                },
                widgets: [
                    'label',
                    'watch',
                    'match',
                    'className',
                    'classes.card',
                    'classes.default.widget',
                    'classes.default.label',
                    'classes.default.field',
                    'classes.default.input'
                ]
            }
        },
        onDropdown: () => Promise.resolve({})
    }
};

const Inspector: ComponentProps = ({
    Editor,
    className,
    object = {},
    type = 'schema',
    property,
    onChange
}) => {
    const classes = useStyles();
    const handleChange = React.useCallback(({$original, ...value}) => {
        onChange((prev = {}) => ({
            ...prev,
            [type]: {...lodashSet(prev[type], property, value)}
        }));
    }, [property, type, onChange]);
    const init = React.useMemo(() => {
        return property ? lodashGet(object, property) : object;
    }, [object, property]);
    return <div className={clsx(className, classes.inspector)}>
        <Editor
            init={init}
            noScroll
            key={String(property)}
            onChange={handleChange}
            {...props[type]}
        />
    </div>;
};

export default Inspector;
