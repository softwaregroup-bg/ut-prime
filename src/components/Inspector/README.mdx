# Inspector
import { Canvas, Story, ArgsTable } from "@storybook/addon-docs";
import {Input, Table} from "./Inspector.stories";

## Importing

```jsx
import Inspector from 'ut-prime/core/Inspector';

<Inspector className='name' />;
```

## Basic Usage

<Canvas>
    <Story id="inspector--basic" />
</Canvas>

## Properties

<ArgsTable story="."/>
