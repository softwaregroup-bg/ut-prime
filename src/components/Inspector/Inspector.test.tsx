import React from 'react';

import { render } from '../test';
import { Basic } from './Inspector.stories';

describe('<Inspector />', () => {
    it('Basic render equals snapshot', async() => {
        const { findByTestId } = render(<Basic {...Basic.args} />);
        expect(await findByTestId('ut-front-test')).toMatchSnapshot();
    });
});
