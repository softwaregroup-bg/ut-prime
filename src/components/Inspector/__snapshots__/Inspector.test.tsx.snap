// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Inspector /> Basic render equals snapshot 1`] = `
<div
  data-testid="ut-front-test"
>
  <div
    class="col-3 inspector-0-2-1"
  >
    <div
      class="flex overflow-x-hidden w-full"
    >
      <div
        class="flex flex-grow-1"
      >
        <div
          class="p-toast p-component p-toast-top-right"
        >
          <div />
        </div>
        <div
          class="grid col align-self-start form-0-2-2"
        >
          <div
            class="col-12 w-full p-0"
          >
            <div
              class="w-full"
            >
              <div
                class="p-card p-component card card-0-2-3"
                id="edit"
              >
                <div
                  class="p-card-body"
                >
                  <div
                    class="p-card-title"
                  >
                    Input
                  </div>
                  <div
                    class="p-card-content"
                  >
                    <div
                      class=""
                    >
                      <div
                        class="field grid"
                      >
                        <label
                          class="col-12 md:col-3"
                          for="title"
                        >
                          Title
                        </label>
                        <div
                          class="flex align-items-center relative col-12 md:col-9"
                        >
                          <input
                            class="p-inputtext p-component w-full"
                            id="title"
                            name="title"
                            value=""
                          />
                        </div>
                      </div>
                      <div
                        class="field grid"
                      >
                        <label
                          class="col-12 md:col-3"
                          for="type"
                        >
                          Type
                        </label>
                        <div
                          class="flex align-items-center relative col-12 md:col-9"
                        >
                          <div
                            class="p-dropdown p-component p-inputwrapper p-dropdown-clearable w-full"
                            data-testid="type"
                            id="type"
                          >
                            <div
                              class="p-hidden-accessible"
                            >
                              <input
                                aria-haspopup="listbox"
                                readonly=""
                                type="text"
                              />
                            </div>
                            <div
                              class="p-hidden-accessible p-dropdown-hidden-select"
                            >
                              <select
                                aria-hidden="true"
                                name="type"
                                tabindex="-1"
                              >
                                <option
                                  selected=""
                                  value=""
                                />
                              </select>
                            </div>
                            <span
                              class="p-dropdown-label p-inputtext p-dropdown-label-empty"
                            >
                              empty
                            </span>
                            <div
                              aria-expanded="false"
                              aria-haspopup="listbox"
                              class="p-dropdown-trigger"
                              role="button"
                            >
                              <span
                                class="p-dropdown-trigger-icon p-clickable pi pi-chevron-down"
                              />
                            </div>
                          </div>
                        </div>
                      </div>
                      <div
                        class="field grid"
                      >
                        <label
                          class="col-12 md:col-3"
                          for="widget-type"
                        >
                          Widget
                        </label>
                        <div
                          class="flex align-items-center relative col-12 md:col-9"
                        >
                          <div
                            class="p-dropdown p-component p-inputwrapper p-dropdown-clearable w-full"
                            data-testid="widget-type"
                            id="widget-type"
                          >
                            <div
                              class="p-hidden-accessible"
                            >
                              <input
                                aria-haspopup="listbox"
                                readonly=""
                                type="text"
                              />
                            </div>
                            <div
                              class="p-hidden-accessible p-dropdown-hidden-select"
                            >
                              <select
                                aria-hidden="true"
                                name="widget.type"
                                tabindex="-1"
                              >
                                <option
                                  selected=""
                                  value=""
                                />
                              </select>
                            </div>
                            <span
                              class="p-dropdown-label p-inputtext p-dropdown-label-empty"
                            >
                              empty
                            </span>
                            <div
                              aria-expanded="false"
                              aria-haspopup="listbox"
                              class="p-dropdown-trigger"
                              role="button"
                            >
                              <span
                                class="p-dropdown-trigger-icon p-clickable pi pi-chevron-down"
                              />
                            </div>
                          </div>
                        </div>
                      </div>
                      <div
                        class="field grid"
                      >
                        <label
                          class="col-12 md:col-3"
                          for="widget-parent"
                        >
                          Parent
                        </label>
                        <div
                          class="flex align-items-center relative col-12 md:col-9"
                        >
                          <input
                            class="p-inputtext p-component w-full"
                            id="widget-parent"
                            name="widget.parent"
                            value=""
                          />
                        </div>
                      </div>
                      <div
                        class="field grid"
                      >
                        <label
                          class="col-12 md:col-3"
                          for="widget-labelClass"
                        >
                          Label Class
                        </label>
                        <div
                          class="flex align-items-center relative col-12 md:col-9"
                        >
                          <div
                            class="p-chips p-component p-inputwrapper p-inputwrapper-filled h-10rem"
                            id="widget-labelClass"
                          >
                            <ul
                              class="p-inputtext p-chips-multiple-container"
                            >
                              <li
                                class="p-chips-token p-highlight"
                              >
                                <span
                                  class="p-chips-token-label"
                                >
                                  md:col-4
                                </span>
                                <span
                                  class="p-chips-token-icon pi pi-times-circle"
                                />
                              </li>
                              <li
                                class="p-chips-input-token"
                              >
                                <input
                                  name="widget.labelClass"
                                  type="text"
                                />
                              </li>
                            </ul>
                          </div>
                          <div
                            class="absolute flex right-0 top-0 bottom-0 justify-content-center align-items-center m-2"
                            style="width: 2.375rem;"
                          >
                            <i
                              class="pi cursor-pointer pi-replay"
                            />
                          </div>
                        </div>
                      </div>
                      <div
                        class="field grid"
                      >
                        <label
                          class="col-12 md:col-3"
                          for="widget-fieldClass"
                        >
                          Field Class
                        </label>
                        <div
                          class="flex align-items-center relative col-12 md:col-9"
                        >
                          <div
                            class="p-chips p-component p-inputwrapper p-inputwrapper-filled h-10rem"
                            id="widget-fieldClass"
                          >
                            <ul
                              class="p-inputtext p-chips-multiple-container"
                            >
                              <li
                                class="p-chips-token p-highlight"
                              >
                                <span
                                  class="p-chips-token-label"
                                >
                                  md:col-8
                                </span>
                                <span
                                  class="p-chips-token-icon pi pi-times-circle"
                                />
                              </li>
                              <li
                                class="p-chips-input-token"
                              >
                                <input
                                  name="widget.fieldClass"
                                  type="text"
                                />
                              </li>
                            </ul>
                          </div>
                          <div
                            class="absolute flex right-0 top-0 bottom-0 justify-content-center align-items-center m-2"
                            style="width: 2.375rem;"
                          >
                            <i
                              class="pi cursor-pointer pi-replay"
                            />
                          </div>
                        </div>
                      </div>
                      <div
                        class="field grid mb-0"
                      >
                        <label
                          class="col-12 md:col-3"
                          for="widget-className"
                        >
                          Input Class
                        </label>
                        <div
                          class="flex align-items-center relative col-12 md:col-9"
                        >
                          <div
                            class="p-chips p-component p-inputwrapper p-inputwrapper-filled h-10rem"
                            id="widget-className"
                          >
                            <ul
                              class="p-inputtext p-chips-multiple-container"
                            >
                              <li
                                class="p-chips-token p-highlight"
                              >
                                <span
                                  class="p-chips-token-label"
                                >
                                  w-full
                                </span>
                                <span
                                  class="p-chips-token-icon pi pi-times-circle"
                                />
                              </li>
                              <li
                                class="p-chips-input-token"
                              >
                                <input
                                  name="widget.className"
                                  type="text"
                                />
                              </li>
                            </ul>
                          </div>
                          <div
                            class="absolute flex right-0 top-0 bottom-0 justify-content-center align-items-center m-2"
                            style="width: 2.375rem;"
                          >
                            <i
                              class="pi cursor-pointer pi-replay"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <pre>
    {
  "type": "object",
  "properties": {
    "test": {
      "sort": true
    }
  }
}
  </pre>
</div>
`;
